feat: snippet completions for functions

llllvvuu opened the pull request:
resolves #56 

https://github.com/pmizio/typescript-tools.nvim/assets/5601392/4c39ab89-c38b-4d47-acfd-57f76c1f12d4

`tsserver` does not create snippets for functions in the completions list; only upon completion resolve (presumably this is for performance reasons).

`typescript-language-server` deals with this by manually constructing the snippet and tweaking the `inputTextFormat`. We copy that here - as long as `includeCompletionsWithSnippetText = true`.

I chose to do one thing different from`typescript-language-server`: even though we can't put preview all the snippet params in the cmp menu, I still at least add "(...)" so that it visually distinguishes them as snippets. This is especially since [`nvim_cmp` is unable to detect them as snippets](https://github.com/hrsh7th/nvim-cmp/issues/1665).

Note that VSCode's own `completeFunctionCalls` feature is [very rough around the edges](https://github.com/microsoft/TypeScript/issues/51758#issue-1477065873). For example, [it doesn't work on arrow functions](https://github.com/microsoft/TypeScript/issues/51427), which is fundamentally because `tsserver` labels then as variables and so it doesn't make sense for them to be `inputTextFormat = Snippet`.

INFO:
refactor/fix: move default configs to `config.lua` so they can be referenced outside of just `text_document/did_open`
test: test both the `completeFunctionCalls = true` case and `completeFunctionCalls = false` case.

pmizio commented:
Very nice! Thank you for contribution! I looked into code and it looks fine, but can you add type annotations for new functions and ad least one test which contain new snippet response for regression purposes?
In the meantime I test it in my setup.

llllvvuu commented:
@pmizio Sure, no problem, just added. For the tests, this feature actually turned the existing test cases into snippets, so I instead added back test cases for non-snippets.

pmizio commented:
> @pmizio Sure, no problem, just added. For the tests, this feature actually turned the existing test cases into snippets, so I instead added back test cases for non-snippets.

I mean more to assert constructed snippet which you add to resolve response [here](https://github.com/pmizio/typescript-tools.nvim/pull/116/files#diff-bb318b020a15617a49a83e92a30945b0ee02d37c1a9129c2768ade494dd91e03R108)

llllvvuu commented:
> I mean more to assert constructed snippet which you add to resolve response [here](https://github.com/pmizio/typescript-tools.nvim/pull/116/files#diff-bb318b020a15617a49a83e92a30945b0ee02d37c1a9129c2768ade494dd91e03R108)

The value is the same as `item.insertText`: https://github.com/pmizio/typescript-tools.nvim/pull/116/files#diff-bb318b020a15617a49a83e92a30945b0ee02d37c1a9129c2768ade494dd91e03R105-R108

Which is tested here:

https://github.com/pmizio/typescript-tools.nvim/pull/116/files#diff-583bd8b57593fbf434045965a180c48661e07c2abd34cbf8c88da15d14e05ad9R197

I'm actually not sure how to trigger the `item.textEdit.newText` route specifically. It comes from `item.replacementSpan` being defined, which doesn't seem to be happening in any case for me.

pmizio commented:
> > I mean more to assert constructed snippet which you add to resolve response [here](https://github.com/pmizio/typescript-tools.nvim/pull/116/files#diff-bb318b020a15617a49a83e92a30945b0ee02d37c1a9129c2768ade494dd91e03R108)
> 
> The value is the same as `item.insertText`: https://github.com/pmizio/typescript-tools.nvim/pull/116/files#diff-bb318b020a15617a49a83e92a30945b0ee02d37c1a9129c2768ade494dd91e03R105-R108
> 
> Which is tested here:
> 
> https://github.com/pmizio/typescript-tools.nvim/pull/116/files#diff-583bd8b57593fbf434045965a180c48661e07c2abd34cbf8c88da15d14e05ad9R197
> 
> I'm actually not sure how to trigger the `item.textEdit.newText` route specifically. It comes from `item.replacementSpan` being defined, which doesn't seem to be happening in any case for me.

Oh, I'm blind sorry :D ok I will work with this for one or two days and merge it if everything will be ok. Once again thanks for contrib!

llllvvuu commented:
@pmizio good catch, I didn't realize it was `nvim-cmp` thing, so I removed the code that adds `~` on our side. Instead I added indicator `(...)`. Unfortunately `nvim-cmp` logic doesn't allow me to trigger the `~` indicator on their side, but I opened an issue for this: https://github.com/hrsh7th/nvim-cmp/issues/1665

I changed one more thing, which is to turn off the feature by default, [similar to VSCode](https://github.com/microsoft/vscode/blob/885dba39a5f546c5077196a8c31b03d03a293b31/extensions/typescript-language-features/package.json#L247-L252). I updated the README.md to show how to turn it on.

pmizio commented on README.md:
```diff
@@ -138,6 +138,9 @@ require("typescript-tools").setup {
     tsserver_file_preferences = {},
+    -- mirror of https://github.com/microsoft/vscode/blob/885dba39a5f546c5077196a8c31b03d03a293b31/extensions/typescript-language-features/package.json#L145-L1266
+    -- `typescript-tools.nvim` currently only supports `typescript.suggest.completeFunctionCalls`
+    vscode_configuration = {},
```
I don't think it need to be called exactly like in vscode, I would prefer simple:
```lua
-- mirror of  vscode `typescript.suggest.completeFunctionCalls` option
-- maybe some url to vscode option docs
complete_function_calls = false,
```

We probably never will support all of vscode options simply because a lot of them are just proxies to `tsserver_format_options` and `tsserver_file_preferences` which are exposed now, so I prefer cherrypicking stuff.

llllvvuu replied:
Sounds good, applied.

pmizio commented on lua/typescript-tools/protocol/text_document/completion/init.lua:
```diff
@@ -65,14 +67,19 @@ function M.handler(request, response, params)
+      local should_create_function_snippet = utils.should_create_function_snippet(kind, filetype)
+      local should_create_snippet = item.isSnippet or should_create_function_snippet
+      local label = is_optional and (item.name .. "?") or item.name
+      label = should_create_function_snippet and (label .. "(...)") or label
```
I'm not sure I like this indicator, it clutters results too much for me. But, it is ok for now especially this feature is disabled by default. If cmp fixes its things can we go back only to tilda?

pmizio replied:
Like I said it is ok for now, maybe cmp fixes this and we just fallback on `~`.

pmizio commented:
Ok, so I tested this and it looks fine for me to merge, @llllvvuu do you want to change something before merge or I can do that?

llllvvuu commented:
@pmizio All good on my end 👍

pmizio commented:
Once again thanks for contrib @llllvvuu!
